src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -303,6 +304,12 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
                             Error(diagnostics, ErrorCode.ERR_MustDeclareForeachIteration, variables);
                             hasErrors = true;
                         }
+                        else
+                        {
+                            // Bind the expression for error recovery, but discard all new diagnostics
+                            iterationErrorExpression = originalBinder.BindExpression(node.Variable, new DiagnosticBag());
```

Comment:
>iterationErrorExpression = originalBinder.BindExpression(node.Variable, new DiagnosticBag()); [](start = 28, length = 93)

It doesn't feel right that we are skipping previous ```else if (!node.HasErrors)```. I think the rule we should follow is simple, the expression should always be bound one way or another. Either we BindDeconstruction, or we BindExpression. There should be no way to skip binding.